dockstore-webservice/src/main/java/io/dockstore/webservice/languages/CWLHandler.java

```diff
@@ -175,10 +178,82 @@ public Version parseWorkflowContent(String filePath, String content, Set<SourceF
+     * language versions of the specified Version.
+     *
+     * In CWL, each SourceFile can contain multiple tools or subworkflows,
+     * each with its own version.  If a SourceFile specifies multiple CWL
+     * versions, the language version is set to the "latest" version per
+     * the list at the bottom of:
+     * https://www.commonwl.org/v1.2/CommandLineTool.html#CWLVersion
+     *
+     * The list of language versions for a given Version is sorted
+     * in descending order, from "latest" to "earliest".
+     */
+    private void setCwlVersionsFromSourceFiles(Set<SourceFile> sourceFiles, Version version) {
+        Set<String> allVersions = new HashSet<>();
+        for (SourceFile file: sourceFiles) {
+            if (file.getType() == DescriptorLanguage.FileType.DOCKSTORE_CWL) {
+                Set<String> fileVersions = getCwlVersionsFromSourceFile(file);
+                file.setTypeVersion(newestVersion(fileVersions));
+                allVersions.addAll(fileVersions);
+            }
+        }
+        version.setDescriptorTypeVersions(sortVersionsDescending(allVersions));
+    }
+
+    /**
+     * Extract the set of CWL language versions that appear in the
+     * specified SourceFile.
+     */
+    private Set<String> getCwlVersionsFromSourceFile(SourceFile file) {
+        Set<String> versions = new HashSet<>();
+        getCwlVersionsFromMap(versions, parseAsMap(file.getContent()));
+        return versions;
+    }
+
+    /**
+     * Recursively examine the parsed YAML Map/List representation of
+     * a CWL file, interpreting the string value of any Map entry with
+     * the key "cwlVersion" as a language version.
+     */
+    private void getCwlVersionsFromMap(Set<String> versions, Map<String, Object> map) {
+        map.forEach((k, v) -> {
+            if (Objects.equals(k, "cwlVersion") && v instanceof String) {
+                versions.add(v.toString());
+            } else if (v instanceof Map) {
+                getCwlVersionsFromMap(versions, (Map<String, Object>)v);
+            }
+        });
+    }
+
+    private String newestVersion(Set<String> versions) {
+        return sortVersionsDescending(versions).stream().findFirst().orElse(null);
+    }
+
+    /**
+     * Sort the specified set of versions in descending order, from
+     * "latest" to "earliest", per the list at the bottom of:
+     * https://www.commonwl.org/v1.2/CommandLineTool.html#CWLVersion
+     *
+     * Since the pattern of CWL versions is to suffix the version number
+     * for development releases, and then use the unsuffixed version number
+     * for the "production" release, we generate sort keys by appending the
+     * maximum value Unicode character to each version.  We also strip off
+     * any "sbg:" prefix, because files with CWL versions like "sbg:draft-2"
+     * are known to exist.
+     */
+    private List<String> sortVersionsDescending(Set<String> versions) {
```

Comment:
Think we have some code for handling versioning using a semantic versioning aware library 
https://github.com/dockstore/dockstore/blob/72a4487e0fe9388058ec8488665acc7438f06e50/dockstore-webservice/src/main/java/io/dockstore/webservice/languages/WDLHandler.java#L560-L582

https://www.commonwl.org/user_guide/introduction/basic-concepts.html#the-cwl-specification (cwl should be compatible)

Reply:
The list of valid CWL versions includes enough irregularities that we might be better off sticking with the PR solution: https://www.commonwl.org/v1.2/CommandLineTool.html#CWLVersion

Reply:
I think you can treat them as validation errors, even the cwl reference runner doesn't support most of them 
https://github.com/common-workflow-language/cwltool/blob/3.1.20221201130942/cwltool/load_tool.py#L441-L453

Reply:
I incorporated the above semver lib into the sorting code.